README.md
armourd
=======

> A Linux daemon for process monitoring targeted at embedded systems running GNU/Linux.

armourd restarts processes that stopped abnormally, by returning a nonzero exit
code. To watch an application armourd only needs the file path of the application
executable; no PID files, *no shell scripts*.

It works out of the box, as it takes roughly a couple of minutes to set it up.
Upon startup armourd interprets its configuration file,
which lists the pathnames of the services that need to be watched, e.g.:

```sh
$ echo '/opt/foo' > /etc/armourd.conf
$ echo '/opt/bar' >> /etc/armourd.conf
```
You can also use file-name globs ('*' wildcard). For example, to watch all the
binaries under /usr/local/bin/:

```sh
$ echo '/usr/local/bin/*' > /etc/armourd.conf
```

See [INSTALL](INSTALL) for details on how to build and run.

Goals
-----

* simplest possible configuration
* no library dependencies (other than the C library)
* small memory footprint
* fast recovery of processes

Caveats
-------

armourd is Linux-only by design, because it relies upon interfaces such as
[epoll(7)](http://man7.org/linux/man-pages/man7/epoll.7.html) or [netlink(7)](http://man7.org/linux/man-pages/man7/netlink.7.html) 
and [proc(5)](http://man7.org/linux/man-pages/man5/proc.5.html) 
files that aren't available in other unices.
It is more suited to sysvinit-based systems; newer init systems, such as
[upstart](http://upstart.ubuntu.com/) or [systemd](http://www.freedesktop.org/wiki/Software/systemd/), 
service recovery is readily available.

armourd does not start services, i.e. its not an init daemon; on the other hand, 
it runs independently of your init system, and thus is portable across Linux devices.

Forking servers will have their children terminate before restart. Although it is 
designed to recover system daemon programs, it can actually recover any application, 
with the limitation that currently only supports single-instance applications 
unless using the [D-Bus](http://dbus.freedesktop.org/) interface (see below).

*It does not recover applications hanging*. 

Dependencies
------------

The minimum required versions are:

* Linux kernel 2.6.27
* GNU C Library (glibc) 2.8
* GCC 4.1
* Autoconf 2.50
* D-Bus 1.6.x (not required)

System Recovery
---------------

armourd can also be configured to talk to watchdog hardware.  In order to 
make your system fully reliable, you should have a process feeding a watchdog, 
although your mileage may vary.  Linux provides a very simple interface to watchdog timers
(`/dev/watchdog`) that armourd implements as a non-default feature, to
provide an all-round solution. To enable it:
```
./configure --enable-watchdog
```

D-Bus Interface
---------------

Optionally, you can build the daemon with D-Bus support to commuincate with other 
applications. The D-Bus API exposes runtime information, and a method to watch 
running processes. The introspection XML data is:

```xml
<node>
  <interface name="org.freedesktop.DBus.Introspectable">
    <method name="Introspect">
      <arg name="xml_data" type="s" direction="out"/>
  </method>
  </interface>
  <interface name="com.github.Armourd">
    <method name="ListProcesses">
      <arg type="a{sv}" direction="out"/>
    </method>
    <method name="WatchProcess">
      <arg name="pid" type="i" direction="in"/>
      <arg name="resultcode" type="u" direction="out"/>
    </method>
  </interface>
</node>
```

Examples
--------

Explicitly request to watch a running process, using
DBus, e.g. with `dbus-send` you would:

```sh
$ dbus-send --session --print-reply --type=method_call \
> --dest=com.github.Armourd /com/github/armourd \
> com.github.Armourd.WatchProcess int32:`pidof foo`
```
Or, using the [example](src/dbus/client-example.py) Python script:

```sh
$ ./client-example.py -w`pidof foo`
```


See http://www.linuxfromscratch.org/blfs/view/svn/general/dbus.html for an example of how to build D-Bus

Testsuite
---------

Current version is not production ready, as there aren't any tests.  Unit tests
to exercise the API's and integration tests to check the runtime behaviour are
a must.  This could be added as a configure option (e.g. --enable-tests) then
the user can run the make *check* target

TODO
----

* Testsuite
* Logging
* Configuration file parser needs to be rewritten (it is buggy)
* Enable options in the configuration file and DBus interface to tailor
  specific needs, e.g. return codes for success, recover script, etc.
* Send a D-Bus signal when a watched application crashes/restarts


debian/changelog
armourd (1.0-local) UNRELEASED; urgency=low

  * Initial release. (Closes: #XXXXXX)

 -- Alex Moreno <alexmorenocano@gmail.com>  Mon, 10 Mar 2014 16:03:51 +0800


src/main.c
/* vim: set sw=4 ts=4 et : */
/* main.c: main() for armourd
 *
 * Copyright (C) 2012 Alexandre Moreno
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 */
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <err.h>

#include "config.h"
#include "armour.h"
#include "armour_dbus.h"
#include "log.h"

static int daemon_mode = 0;

#ifdef SYSCONFDIR
static char *config_file = SYSCONFDIR "/armourd.conf";
#else
static char *config_file = "/etc/armourd.conf";
#endif

static int running (void)
{
    int fd;
    struct flock fl;
    char buf[16];

    if ((fd = open ("/var/run/armourd.pid"
        , O_RDWR | O_CREAT, 0644)) < 0) {
	    err (EXIT_FAILURE, "open");
    }

    fl = (struct flock) { .l_type = F_WRLCK, .l_whence = SEEK_SET };
    
    if (fcntl (fd, F_SETLK, &fl) < 0) {
        if (EACCES == errno || EAGAIN == errno) {
            close (fd);
            return 1;
        }
	    err (EXIT_FAILURE, "fcntl");
    }

    fcntl (fd, F_SETFD, FD_CLOEXEC);

    ftruncate (fd, 0);
    sprintf (buf, "%ld", (long)getpid());
    write (fd, buf, strlen (buf) + 1);

    /* keep it open to hold the lock */
    return 0;
}

static void usage (void)
{
    fprintf (stderr, "%s [--daemon] [--version] [--config-file=FILE]\n"
            , PACKAGE_NAME);
    exit (0);
}

static void version (void)
{
    printf ("%s\n"
            "Copyright (C) 2013, Alexandre Moreno\n"
            "This is free software: you are free to change and redistribute it.\n"
            "There is NO WARRANTY, to the extent permitted by law.\n"
            , PACKAGE_STRING);
    exit (0);
}

static void get_options (int argc, char **argv)
{
    /* parse command-line options */
    for (;;) {
        static const struct option long_options[] = {
            {"daemon", 0, NULL, 'd'},
            {"config-file", 1, NULL, 'c'},
            {"verbose" , 0, NULL, 'v'},
            {"help" , 0, NULL, 'h'},
            {"version" , 0, NULL, 'V'},
            {0, 0, 0, 0},
        };

        int c = getopt_long (argc, argv, "c:dhv", long_options, NULL);
        if (c == -1)
            break;

        switch (c) {
            case 'd':
                daemon_mode = 1;
                enable_syslog("armourd");
                break;
            case 'v':
                _verbose = 1;
                break;
            case 'c':
                config_file = strdup (optarg);
                break;
            case 'V':
                version ();
                break;
            case 'h':
            case '?':
            default:
                usage ();
                break;
        }
    }
}

int main (int argc, char **argv)
{
    armour_t *self;

    get_options (argc, argv);

    if (daemon_mode
    && (daemon (0, 1) < 0)) {
        err (EXIT_FAILURE, "daemon()");
    }

    if (running ()) {
        errx (EXIT_FAILURE, "already running");
    }

    if (armour_init (&self, config_file) < 0) {
        errx (EXIT_FAILURE, "failed to start");
    }

    if (armour_dbus_init (self) < 0) {
        warn ("dbus interface init failed");
    }

	if (armour_watchdog_init (&self->watchdog) < 0) {
		warn ("watchdog interface init failed");
	}

    return armour_run (self);
}
